### DISCUSSION TITLE
Unexpected unicode character translation in math mode

### DISCUSSION THREAD

**Discussion by kunzaatko** [typst/typst#4167] [General]
### Description

Copying the unicode character of a [roundhand caligraphic D](https://en.wikipedia.org/wiki/Mathematical_Alphanumeric_Symbols#Glyph_variants) 𝒟 (U+1D49F) into normal mode text yields the correct character:
![image](https://github.com/typst/typst/assets/56647779/040a586e-ed2d-4831-814c-13dc8d5c9b92) 
![image](https://github.com/typst/typst/assets/56647779/a6449941-ca0c-4107-a2f6-35d8f2c92cb7)

However if it is in math mode, the character is translated into chancery caligraphic.
![image](https://github.com/typst/typst/assets/56647779/e4f5bf5b-f132-4074-8d70-177caed9db76)
![image](https://github.com/typst/typst/assets/56647779/f78a5f8d-b45e-4738-bd3a-2d965da50609)

Presumably, this is not isolated to D, but this was not tested.

This issue is probably located here: [math styling file](https://github.com/typst/typst/blob/aad68ea48972787ca540295c4ee740aec6a72a0b/crates/typst/src/math/style.rs#L292), and should be possible to solve by mapping the unicode table roundhand math variant from the [wikipedia page](https://en.wikipedia.org/wiki/Mathematical_Alphanumeric_Symbols#Glyph_variants) in the file.

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by kunzaatko** [comment]
Possibly related #4139

**Comment by kunzaatko** [comment]
Same issue occurs even when using the function `from-unicode`
![image](https://github.com/typst/typst/assets/56647779/2d2c1791-040a-41fb-9e33-28d0ec3e246c)
![image](https://github.com/typst/typst/assets/56647779/156a8929-e21c-4ca3-9061-f78ffdd7e863)

**Comment by Enivex** [comment]
This is not a bug. Chancery and roundhand occupy the same code points in unicode, so it's entirely up to the font which one they pick. Some fonts have both, hidden behind stylistic sets https://typst.app/docs/reference/text/text/#parameters-stylistic-set, but there's no real consistency between exactly which one.

**Comment by kunzaatko** [comment]
OK, thanks for the explanation. However I would argue, that there is no reason to introduce this inconsistency between normal and math mode. Why do they use a different stylistic set by default? I didn't change any font setting in the document otherwise.. Is this also the desired functionality?

**Comment by kunzaatko** [comment]
Furthermore, the stylistic sets are not working for me in both normal and math mode. I am using New Computer Modern, where in the [documentation](https://mirrors.nic.cz/tex-archive/fonts/newcomputermodern/doc/newcm-doc.pdf) on page 18 it is stated that the roundhand is located in stylistic set 1 and the default should be chancery. However in normal mode
```typst
#text(stylistic-set: none)[#str.from-unicode(119967)]
```
still gives 
![image](https://github.com/typst/typst/assets/56647779/e82c8d5c-6051-4458-ba18-fb01427a8238)

In math mode
```
$#text( stylistic-set: 1)[#str.from-unicode(119967)]$
```
gives 
![image](https://github.com/typst/typst/assets/56647779/7c723a1f-89b6-48b1-b832-5c27fa85f991).
Is this be a bug or am I missing anything?

**Comment by Enivex** [comment]
> OK, thanks for the explanation. However I would argue, that there is no reason to introduce this inconsistency between normal and math mode. Why do they use a different stylistic set by default? I didn't change any font setting in the document otherwise.. Is this also the desired functionality?

There is no inconsistency other than the text (Linux Libertine) and math (New Computer Modern Math) fonts making different choices.

There is also the possibility of fallback (I haven't checked), since text fonts rarely include this glyph.

**Comment by Enivex** [comment]
> Furthermore, the stylistic sets are not working for me in both normal and math mode. I am using New Computer Modern, where in the [documentation](https://mirrors.nic.cz/tex-archive/fonts/newcomputermodern/doc/newcm-doc.pdf) on page 18 it is stated that the roundhand is located in stylistic set 1 and the default should be chancery. However in normal mode
> ```typst
> #text(stylistic-set: none)[#str.from-unicode(119967)]
> ```
> still gives 
> ![image](https://github.com/typst/typst/assets/56647779/e82c8d5c-6051-4458-ba18-fb01427a8238)
> 
> In math mode
> ```
> $#text( stylistic-set: 1)[#str.from-unicode(119967)]$
> ```
> gives 
> ![image](https://github.com/typst/typst/assets/56647779/7c723a1f-89b6-48b1-b832-5c27fa85f991).
> Is this be a bug or am I missing anything?

The documentation is talking about New Computer Modern **Math** specifically. There's no reason to expect the stylistic set works for the text font, especially since it's not even New Computer Modern by default. In fact, it would surprise me if the text font even includes that glyph (like I mentioned above).

In general this kind of symbol is only supposed to be used in math mode.

**Reply by kunzaatko** [reply]
Yeah, sorry. That is actually the font that I am using _New Computer Modern Math_
```typst
#show math.equation: set text(font: "New Computer Modern Math")
```
Is there a reason, why the setting of `stylistic-set: 1` has no effect in math mode, then?
```typst
$#text( stylistic-set: 1)[#str.from-unicode(119967)]$
```
![image](https://github.com/typst/typst/assets/56647779/243eb46d-6695-428d-921d-5d47fbe90392)

**Reply by kunzaatko** [reply]
You where right with the normal mode though. I forgot to set the normal mode font to _New Computer Modern_ for the test. I set it up now with a _New Computer Modern Math_ as a fallback and it works as expected.
```typst
#set text(font: ("New Computer Modern", "New Computer Modern Math"))
```
with
```typst
#text( stylistic-set: 1)[#str.from-unicode(119967)]#text( stylistic-set: none)[#str.from-unicode(119967)]
```
yeilding 
![image](https://github.com/typst/typst/assets/56647779/7e9f0b83-fb61-4fc3-bc21-686ca05ed23f)

**Reply by Enivex** [reply]
> Yeah, sorry. That is actually the font that I am using _New Computer Modern Math_
> ```typst
> #show math.equation: set text(font: "New Computer Modern Math")
> ```
> Is there a reason, why the setting of `stylistic-set: 1` has no effect in math mode, then?
> ```typst
> $#text( stylistic-set: 1)[#str.from-unicode(119967)]$
> ```
> ![image](https://github.com/typst/typst/assets/56647779/243eb46d-6695-428d-921d-5d47fbe90392)

Likely running into this issue https://github.com/typst/typst/issues/3379

**Reply by kunzaatko** [reply]
Yeah, I bet that is it. Thank you very much for your support!
